[PGNCCL] Slimming watchdog loop

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139834

- Refactored traceback code into `work.printTraceback()`.  cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @shuqiangzhang
- Refactored desync debug code into `class DesyncDebugger`.
- Moved occurrences of `futureWorkResult_->markCompleted` into `checkAndSetException` and `checkTimeout`, respectively. cc @shuqiangzhang 
- Modularized dump signal broadcast code into `ProcessGroupNCCL::broadcastDumpSignal`. cc @fduwjj @c-p-i-o 

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

kwen2501 committed:
[PGNCCL] Slimming watchdog loop

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Slimming watchdog loop"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Slimming watchdog loop"


- Refactored traceback code into `work.printTraceback()`.  cc shuqiangzhang
- Refactored desync debug code into `class DesyncDebugger`.
- Moved occurrences of `futureWorkResult_->markCompleted` into `checkAndSetException` and `checkTimeout`, respectively. cc shuqiangzhang 
- Modularized dump signal broadcast code into `ProcessGroupNCCL::broadcastDumpSignal`. cc fduwjj c-p-i-o 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Slimming watchdog loop"


- Refactored traceback code into `work.printTraceback()`.  cc shuqiangzhang
- Refactored desync debug code into `class DesyncDebugger`.
- Moved occurrences of `futureWorkResult_->markCompleted` into `checkAndSetException` and `checkTimeout`, respectively. cc shuqiangzhang 
- Modularized dump signal broadcast code into `ProcessGroupNCCL::broadcastDumpSignal`. cc fduwjj c-p-i-o 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 committed:
Update on "[PGNCCL] Slimming watchdog loop"


- Refactored traceback code into `work.printTraceback()`.  cc shuqiangzhang
- Refactored desync debug code into `class DesyncDebugger`.
- Moved occurrences of `futureWorkResult_->markCompleted` into `checkAndSetException` and `checkTimeout`, respectively. cc shuqiangzhang 
- Modularized dump signal broadcast code into `ProcessGroupNCCL::broadcastDumpSignal`. cc fduwjj c-p-i-o 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1977,91 +2058,34 @@ void ProcessGroupNCCL::watchdogHandler() {
         work.checkAndSetException();
       }
+      // Then check if work has timed out
+      bool timedout = work.checkTimeout();
```
This basically changes the existing error logic which intentionally differentiate nccl errors from timeout. There is benefit of doing so that we can identify the type of error immediately from the LOGs.

kwen2501 replied:
In `checkTimeout`, exception will not be changed if it has been set.

shuqiangzhang replied:
Understood. Not just the exception logging, I was talking about the logging within the `if (work.exception()) {} `branch. Previously, it would be logged only for nccl errors, and only if there is no nccl errors, we checked the timeout, and timeout error would have a different logging. It's very handy in debugging when we only log the first/true errors.  If we found an nccl error, there is no need to check the timeout. that's why I think the existing if/else structure is preferred.

kwen2501 replied:
Got it. I believe `checkAndSetException` has logging too?

kwen2501 replied:
Yeah, I checked, it has logging.

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1977,91 +2058,34 @@ void ProcessGroupNCCL::watchdogHandler() {
+        desyncDebugger_.run();
+      }
+
+      // If work hits an exception (either an error or timeout)
```
This would make it harder to debug whether the execption is nccl error or timeout, We'd better clearly differentiate the two types of logs, because usually there are triggered by very different reasons.

kwen2501 replied:
This is old code. I didn't change the behavior. Can we address in a new PR if needed?

shuqiangzhang commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -691,7 +692,23 @@ bool ProcessGroupNCCL::WorkNCCL::checkTimeout(
+  }
+
+  // Mark future result as TIMEOUT
+  if (futureWorkResult_ && !futureWorkResult_->completed()) {
```
LGTM that we moved this to checktimeout, but one caveat is that checkTimeout is being called not just from watchdog thread, but user thread too (e.g., when user calls work.wait(timeout)), so the callback registered by the user could be executed in the user thread in this case. But I think this should be OK, or even desirable.

kwen2501 replied:
I think it is supported to mark completed *before* a callback is attached to a Future. The callback will directly get executed at attach time in that case.

kwen2501 committed:
Update on "[PGNCCL] Slimming watchdog loop"


- Refactored traceback code into `work.printTraceback()`.  cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o shuqiangzhang
- Refactored desync debug code into `class DesyncDebugger`.
- Moved occurrences of `futureWorkResult_->markCompleted` into `checkAndSetException` and `checkTimeout`, respectively. cc shuqiangzhang 
- Modularized dump signal broadcast code into `ProcessGroupNCCL::broadcastDumpSignal`. cc fduwjj c-p-i-o 

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
